Add ZerosSampler

fedimser opened the pull request:
A dummy sampler which implements `cirq.Sampler` and returns zeroes for all measurements.

Fixes #2818

mpharrigan commented on cirq/work/zeros_sampler.py:
```diff
@@ -0,0 +1,63 @@
+            TrialResult list for this run; one for each possible parameter
+            resolver.
+        """
+        zero_measurement = np.zeros((repetitions, 1), dtype=np.int8)
```
this doesn't work correctly for circuits with multi-qubit measurement gates

fedimser replied:
I didn't know Cirq support that :)
Added support for multi-qubit measurement, PTAL.

mpharrigan commented on cirq/work/zeros_sampler_test.py:
```diff
@@ -0,0 +1,54 @@
+
+def test_sample():
+    # Create a circuit whose measurements are always zeros, and check that
+    # results of ZeroSampler on this circuit are identical to results of
```
👍

mpharrigan commented on cirq/work/zeros_sampler.py:
```diff
@@ -0,0 +1,56 @@
+            resolver.
+        """
+        measurements = {}  # type: Dict[str, np.ndarray]
+        for op in ops.flatten_op_tree(iter(program)):
```
use `program.all_operations`?

fedimser commented:
Hi Matthew,

1. ZeroSampler inherits from Sampler, which has default implementation of `run_async` which  calls `run` and of `run_sweep_async` which calls `run_sweep`. So now async methods of ZerosSampler will work as expected, and I don't need to implement them. Am I right?
2. I didn't work with gate sets in Cirq before, but here is what I can suggest. We can pass `SerializableGateSet` object to ZeroSampler constructor. Then in `run_sweep` we will check for all operations whether it `is_supported_operation`. And if there is no gate set, don't do this validation. Does it make sense? Do you think it's better to do this in a follow-up PR?
Also, what are "extand qubits"?

mpharrigan commented:
1: ok good
2: that sounds entirely reasonable to me
2b, re: follow up PR. up to you. If it's not too drastic of a change, I'd say just put it into this because it will likely involve changing the API of `ZerosSampler` slightly
2c: re: extant qubits. real devices only have certain qubits and a common failure mode is accidentally using e.g. a LineQubit when you should use GridQubits or using GridQubit(0,0) when the physical numbering scheme starts somewhere else

fedimser commented:
Hi @mpharrigan,
I added gateset validation, PTAL.

Regarding extand qubits: there is a class cirq.devices.Device, which has method qubit_set. So we could pass the device to a constructor (like I did with gate set) and in run_sweep() get all qubits referenced by a device, and check that they are all in this qubit_set.
However, this can be simplified: we could just pass qubit_set. Or, if we are passing device, we can just call Device.validate_circuit. 
What do you prefer? 
Also, does Device.validate_circuit validates gateset? If so we don't need SerializableGetSet in ZerosSampler (which I just added).

dabacon commented:
Question to @mpharrigan : shouldn't the device and gateset validation be handled by setting an appropriate device on the circuit?  What is the primary use case for this class, is it to do this validation?  Then maybe it should be called something slightly different?

fedimser commented on cirq/work/zeros_sampler.py:
```diff
@@ -0,0 +1,70 @@
+
+if TYPE_CHECKING:
+    import cirq
+    from cirq.google.serializable_gate_set import SerializableGateSet
```
Done, except now I must violate line length limit.

fedimser commented on cirq/work/zeros_sampler_test.py:
```diff
@@ -0,0 +1,71 @@
+    circuit1 = cirq.Circuit([cirq.X(a)])
+    circuit2 = cirq.Circuit([cirq.CX(a, b)])
+
+    sampler.sample(circuit1)
```
Done.

fedimser commented:
Can we merge this?
@mpharrigan @dabacon

balopat commented:
@googlebot  please rescan
